src/controller/user.ts

```diff
@@ -0,0 +1,214 @@
+        userToBeUpdated.name = context.request.body.name
+        userToBeUpdated.email = context.request.body.email
+        userToBeUpdated.dob = context.request.body.dob
+        userToBeUpdated.address = context.request.body.address
+        userToBeUpdated.description = context.request.body.description
+        delete userToBeUpdated.createdAt
+
+        const validationErrors = await validate(userToBeUpdated, userSchema, ['id'])
+
+        if (validationErrors) return response(context, 400, validationErrors)
+
+        let call = await safeCall(userRepository.findOne(userToBeUpdated.id))
+
+        if (call.err) return response(context, 500, call.err)
+
+        if (!call.res) return response(context, 400, 'user_not_found')
+
+        call = await safeCall(
+            userRepository.findOne({ id: Not(Equal(userToBeUpdated.id)), email: userToBeUpdated.email }),
+        )
+
+        if (call.err) return response(context, 500, call.err)
+
+        if (call.res) return response(context, 400, 'user_already_exists')
+
+        if (context.state.user.email !== userToBeUpdated.email) return response(context, 403, 'not_authorized')
+
+        if (userToBeUpdated.dob) userToBeUpdated.dob = new Date(userToBeUpdated.dob)
+
+        call = await safeCall(userRepository.save(userToBeUpdated))
+
+        if (call.err) return response(context, 500, call.err)
+
+        response(context, 200, call.res.toJSON())
+    }
+
+    @request('delete', '/users/{id}')
+    @summary('Delete user by id')
+    @path({ id: { type: 'string', required: true, description: 'id of user' } })
+    @responses({
+        204: { description: 'user deleted successfully' },
+        400: { description: 'user not found, validation errors' },
+        401: { description: 'token authorization error' },
+        403: { description: 'not authorized to perform this action' },
+    })
+    public static async deleteUser(context: BaseContext): Promise<void> {
+        const validationErrors = await validate({ id: context.params.id }, userSchema, ['id'])
+
+        if (validationErrors) return response(context, 400, validationErrors)
+
+        const userRepository = getManager().getRepository(User)
+        const userToRemove: User | undefined = await userRepository.findOne(context.params.id)
+
+        if (!userToRemove) return response(context, 400, 'user_not_found')
+
+        if (context.state.user.email !== userToRemove.email) return response(context, 403, 'not_authorized')
+
+        const call = await safeCall(userRepository.remove(userToRemove))
+
+        if (call.err) return response(context, 500, call.err)
+
+        response(context, 204)
+    }
+
+    @request('delete', '/testusers')
+    @summary('Delete all test users')
+    @responses({
+        204: { description: 'all users deleted successfully' },
+        401: { description: 'token authorization error' },
+    })
+    public static async deleteTestUsers(context: BaseContext): Promise<void> {
```

Comment:
who will use this API, and how can we prevent other users from using it.

Reply:
For the purposes of this back-end test, I exposed this endpoint to all authenticated test users to allow for testing on the online build. In actual production, I would not expose this kind of sensitive endpoint that can remove all users from the database. However, one has to be authenticated to use it for the test as well. 

In the rare scenario where such an endpoint actually is needed in production, I would restrict it to administrators and protect it with several layers of authentication:
1. Auth token to be authorize user on the server
2.  Check for user role (e.g. Super Admin, Admin, Mod, User, Subscriber). Only one as Super Admin can execute this action.
3. Possibly add another endpoint for Two-Factor Authentication and require this authentication to permit this sensitive admin action.